File: lib/rust/src/engine.rs
//! Radix tree engine implementations with different concurrency models

use std::net::IpAddr;
use std::sync::{
    Arc, RwLock,
    atomic::{AtomicUsize, Ordering},
};
use sysinfo::System;

use crate::traits::*;
use crate::tree::{CompressedTree, UncompressedTree};
use crate::types::{EngineStats, Metadata};
use ipnetwork::IpNetwork;

// ============================================================
// STANDARD ENGINE
// ============================================================

pub struct StandardEngine<T: RouteTree> {
    tree: T,
    size: AtomicUsize,
    stats: RwLock<EngineStats>,
}

impl<T: RouteTree> StandardEngine<T> {
    pub fn new(tree: T) -> Self {
        Self {
            tree,
            size: AtomicUsize::new(0),
            stats: RwLock::new(EngineStats::default()),
        }
    }

    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        let is_new = self.tree.insert(prefix, metadata)?;

        if is_new {
            self.size.fetch_add(1, Ordering::Relaxed);
        }

        let mut stats = self.stats.write().unwrap();
        stats.inserts += 1;
        stats.size = self.size();
        Ok(())
    }

    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        let result = self.tree.lookup(ip);

        let mut stats = self.stats.write().unwrap();
        stats.lookups += 1;
        if result.is_some() {
            stats.hits += 1;
        } else {
            stats.misses += 1;
        }

        result
    }

    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        let removed = self.tree.remove(prefix);

        if removed.is_some() {
            self.size.fetch_sub(1, Ordering::Relaxed);
            let mut stats = self.stats.write().unwrap();
            stats.removals += 1;
            stats.size = self.size();
        }

        removed
    }

    fn contains(&self, prefix: &IpNetwork) -> bool {
        self.tree.contains(prefix)
    }

    fn clear(&self) {
        self.tree.clear();
        self.size.store(0, Ordering::Relaxed);
        let mut stats = self.stats.write().unwrap();
        stats.size = 0;
    }

    fn size(&self) -> usize {
        self.size.load(Ordering::Relaxed)
    }

    fn stats(&self) -> EngineStats {
        let mut stats = self.stats.read().unwrap().clone();
        stats.size = self.size();
        stats
    }
}

impl<T: RouteTree> RadixEngine for StandardEngine<T> {
    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        StandardEngine::insert(self, prefix, metadata)
    }
    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        StandardEngine::lookup(self, ip)
    }
    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        StandardEngine::remove(self, prefix)
    }
    fn contains(&self, prefix: &IpNetwork) -> bool {
        StandardEngine::contains(self, prefix)
    }
    fn clear(&self) {
        StandardEngine::clear(self)
    }
    fn size(&self) -> usize {
        StandardEngine::size(self)
    }
    fn stats(&self) -> EngineStats {
        StandardEngine::stats(self)
    }
}

// SHARDED ENGINE
// throughput = num_shards * throughput_per_shard

pub struct ShardedEngine<T: RouteTree> {
    pub shards: Vec<Arc<StandardEngine<T>>>,
    pub num_shards: usize,
    pub mask_bits: u8,
}

impl<T: RouteTree + Clone> ShardedEngine<T> {
    fn new(num_shards: usize, tree: T) -> Self {
        let shards = (0..num_shards)
            .map(|_| Arc::new(StandardEngine::new(tree.clone())))
            .collect();
        Self {
            shards,
            num_shards,
            mask_bits: 32,
        }
    }

    fn get_shard(&self, ip: &IpAddr) -> usize {
        let hash = match ip {
            IpAddr::V4(ip) => {
                let ip_u32 = u32::from_be_bytes(ip.octets());
                 // Mask the IP to keep only prefix bits
                let mask = if self.mask_bits >= 32 {
                    u32::MAX
                } else {
                    u32::MAX << (32 - self.mask_bits)
                };
                // this way we isolate the ip-adress from the network prefix
                let masked_ip = ip_u32 & mask;

                 // Hash the masked IP
                let mut hash = 0u32;
                for byte in masked_ip.to_be_bytes() {
                    hash = hash.wrapping_mul(31).wrapping_add(byte as u32);
                }
                hash as usize
            }
            IpAddr::V6(ip) => {
                let bytes = ip.octets();
                let mut hash = 0u64;
                if self.mask_bits <= 64 {
                     // Hash only the masked prefix part
                    let bytes_to_keep = ((self.mask_bits + 7) / 8) as usize;
                    for byte in bytes.iter().take(bytes_to_keep) {
                        hash = hash.wrapping_mul(31).wrapping_add(*byte as u64);
                    }
                    // Handle partial byte masking if needed
                    if self.mask_bits % 8 != 0 && bytes_to_keep < bytes.len() {
                        let mask = 0xFFu8 << (8 - (self.mask_bits % 8) as u8);
                        let partial_byte = bytes[bytes_to_keep] & mask;
                        hash = hash.wrapping_mul(31).wrapping_add(partial_byte as u64);
                    }
                } else {
                    // Hash first 16 bytes for large prefix lengths
                    for byte in bytes.iter().take(16) {
                        hash = hash.wrapping_mul(31).wrapping_add(*byte as u64);
                    }
                };
                hash as usize
            }
        };
        hash % self.num_shards
    }
}

impl<T: RouteTree + Clone> RadixEngine for ShardedEngine<T> {
    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        for shard in &self.shards {
            shard.insert(prefix, metadata.clone())?;
        }
        Ok(())
    }

    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        let shard_idx = self.get_shard(ip);
        self.shards[shard_idx].lookup(ip)
    }

    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        let mut removed = None;
        for shard in &self.shards {
            let shard_removed = shard.remove(prefix);
            if removed.is_none() {
                removed = shard_removed;
            }
        }
        removed
    }

    fn contains(&self, prefix: &IpNetwork) -> bool {
        self.shards
            .first()
            .map(|s| s.contains(prefix))
            .unwrap_or(false)
    }

    fn clear(&self) {
        for shard in &self.shards {
            shard.clear();
        }
    }

    fn size(&self) -> usize {
        self.shards.first().map(|s| s.size()).unwrap_or(0)
    }

    fn stats(&self) -> EngineStats {
        let mut total = EngineStats::default();
        for shard in &self.shards {
            let stats = shard.stats();
            total.lookups += stats.lookups;
            total.hits += stats.hits;
            total.misses += stats.misses;
        }
        if let Some(first) = self.shards.first() {
            let stats = first.stats();
            total.inserts = stats.inserts;
            total.removals = stats.removals;
        }
        total.size = self.size();
        total
    }
}

// ENGINE WRAPPER
// switches between concurrency models behind one type

#[derive(Clone)]
pub enum EngineWrapper {
    StandardUncompressed(Arc<StandardEngine<UncompressedTree>>),
    ConcurrentUncompressed(Arc<ShardedEngine<UncompressedTree>>),
    StandardCompressed(Arc<StandardEngine<CompressedTree>>),
    ConcurrentCompressed(Arc<ShardedEngine<CompressedTree>>),
    StandardART(Arc<crate::engine_art::ARTEngineAdapter>),
    ConcurrentART(Arc<crate::engine_art::ShardedARTEngineAdapter>),
}

impl EngineWrapper {
    /// `num_shards`:
    ///   `None`      -> auto-detect from CPU cores + available memory
    ///   `Some(1)`   -> standard (unsharded)
    ///   `Some(n>1)` -> sharded with exactly n shards
    pub fn new(
        variant: EngineVariant,
        node_variant: NodeVariant,
        compressed: bool,
        num_shards: Option<usize>,
    ) -> Self {
        let shard_count = num_shards.unwrap_or_else(|| Self::calculate_optimal_shards(&variant));

        // ART carries its own sharding implementation, handled separately.
        if variant == EngineVariant::ART {
            return if shard_count > 1 {
                EngineWrapper::ConcurrentART(Arc::new(
                    crate::engine_art::ShardedARTEngineAdapter::new(shard_count),
                ))
            } else {
                EngineWrapper::StandardART(Arc::new(crate::engine_art::ARTEngineAdapter::new()))
            };
        }

        // LockFree and Adaptive pick their own node representation;
        // everything else uses whatever node_variant the caller passed.
        let resolved_node_variant = match variant {
            EngineVariant::LockFree if compressed => NodeVariant::LockFreeRadixNode,
            EngineVariant::LockFree => NodeVariant::LockFreeTrieNode,
            EngineVariant::Adaptive if compressed => NodeVariant::AtomicRadixNode,
            EngineVariant::Adaptive => NodeVariant::AtomicTrieNode,
            _ => node_variant,
        };

        // LockFree relies on lock-free nodes for its concurrency, not sharding.
        let effective_shards = match variant {
            EngineVariant::LockFree => 1,
            _ => shard_count,
        };

        if compressed {
            let tree = CompressedTree::new(resolved_node_variant);
            if effective_shards > 1 {
                EngineWrapper::ConcurrentCompressed(Arc::new(ShardedEngine::new(
                    effective_shards,
                    tree,
                )))
            } else {
                EngineWrapper::StandardCompressed(Arc::new(StandardEngine::new(tree)))
            }
        } else {
            let tree = UncompressedTree::new(resolved_node_variant);
            if effective_shards > 1 {
                EngineWrapper::ConcurrentUncompressed(Arc::new(ShardedEngine::new(
                    effective_shards,
                    tree,
                )))
            } else {
                EngineWrapper::StandardUncompressed(Arc::new(StandardEngine::new(tree)))
            }
        }
    }

    // shard-count auto-detection 

    /// Cores -> shards, roughly matching:
    /// Dev(2c)->1-2, Small(4c)->4-8, Medium(8c)->16-32, Large(16c)->32-64, Enterprise(32c)->64-128
    fn calculate_optimal_shards(variant: &EngineVariant) -> usize {
        match variant {
            EngineVariant::Standard | EngineVariant::LockFree => 1,
            EngineVariant::ART => Self::shards_from_cpu(&[(2, 2), (4, 4), (8, 8)], 16),
            EngineVariant::Concurrent => {
                Self::shards_from_cpu(&[(2, 2), (4, 8), (8, 16), (16, 32)], 64)
            }
            EngineVariant::Adaptive => {
                let cpu_based = Self::shards_from_cpu(&[(2, 2), (4, 8), (8, 16), (16, 32)], 64);
                let mem_based = Self::shards_from_memory();
                // Take the more conservative bound so we don't over-shard
                // (and thus over-allocate) on memory-constrained boxes.
                cpu_based.min(mem_based)
            }
        }
    }

    /// `breakpoints` is `[(max_cpus, shards), ...]` ascending; `default`
    /// applies once core count exceeds every breakpoint.
    fn shards_from_cpu(breakpoints: &[(usize, usize)], default: usize) -> usize {
        let cpus = std::thread::available_parallelism()
            .map(|c| c.get())
            .unwrap_or(1);
        breakpoints
            .iter()
            .find(|(max_cpus, _)| cpus <= *max_cpus)
            .map(|(_, shards)| *shards)
            .unwrap_or(default)
    }

    fn shards_from_memory() -> usize {
        const MB: u64 = 1024 * 1024;
        let usable_mb = Self::usable_memory_bytes() / MB;
        match usable_mb {
            0..=512 => 2,
            513..=2048 => 8,
            2049..=4096 => 16,
            4097..=8192 => 32,
            _ => 64,
        }
    }

    /// Only budget 75% of currently *available* RAM — leave headroom for
    /// the OS, allocator, and other processes rather than assuming the
    /// engine owns the whole machine.
    fn usable_memory_bytes() -> u64 {
        Self::detect_available_memory().saturating_mul(3) / 4
    }

    /// Cross-platform available memory via `sysinfo` (Linux/macOS/Windows/BSD/...).
    /// Deliberately uses `available_memory`, not `total_memory` — total tells
    /// you nothing about what's actually free to allocate right now, and
    /// `System::new()` + `refresh_memory()` avoids pulling in CPU/process/disk
    /// data we don't need just to answer this one question.
    fn detect_available_memory() -> u64 {
        let mut system = System::new();
        system.refresh_memory();
        system.available_memory()
    }
}

impl RadixEngine for EngineWrapper {
    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        match self {
            EngineWrapper::StandardUncompressed(e) => e.insert(prefix, metadata),
            EngineWrapper::ConcurrentUncompressed(e) => e.insert(prefix, metadata),
            EngineWrapper::StandardCompressed(e) => e.insert(prefix, metadata),
            EngineWrapper::ConcurrentCompressed(e) => e.insert(prefix, metadata),
            EngineWrapper::StandardART(e) => e.insert(prefix, metadata),
            EngineWrapper::ConcurrentART(e) => e.insert(prefix, metadata),
        }
    }

    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        match self {
            EngineWrapper::StandardUncompressed(e) => e.lookup(ip),
            EngineWrapper::ConcurrentUncompressed(e) => e.lookup(ip),
            EngineWrapper::StandardCompressed(e) => e.lookup(ip),
            EngineWrapper::ConcurrentCompressed(e) => e.lookup(ip),
            EngineWrapper::StandardART(e) => e.lookup(ip),
            EngineWrapper::ConcurrentART(e) => e.lookup(ip),
        }
    }

    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        match self {
            EngineWrapper::StandardUncompressed(e) => e.remove(prefix),
            EngineWrapper::ConcurrentUncompressed(e) => e.remove(prefix),
            EngineWrapper::StandardCompressed(e) => e.remove(prefix),
            EngineWrapper::ConcurrentCompressed(e) => e.remove(prefix),
            EngineWrapper::StandardART(e) => e.remove(prefix),
            EngineWrapper::ConcurrentART(e) => e.remove(prefix),
        }
    }

    fn contains(&self, prefix: &IpNetwork) -> bool {
        match self {
            EngineWrapper::StandardUncompressed(e) => e.contains(prefix),
            EngineWrapper::ConcurrentUncompressed(e) => e.contains(prefix),
            EngineWrapper::StandardCompressed(e) => e.contains(prefix),
            EngineWrapper::ConcurrentCompressed(e) => e.contains(prefix),
            EngineWrapper::StandardART(e) => e.contains(prefix),
            EngineWrapper::ConcurrentART(e) => e.contains(prefix),
        }
    }

    fn clear(&self) {
        match self {
            EngineWrapper::StandardUncompressed(e) => e.clear(),
            EngineWrapper::ConcurrentUncompressed(e) => e.clear(),
            EngineWrapper::StandardCompressed(e) => e.clear(),
            EngineWrapper::ConcurrentCompressed(e) => e.clear(),
            EngineWrapper::StandardART(e) => e.clear(),
            EngineWrapper::ConcurrentART(e) => e.clear(),
        }
    }

    fn size(&self) -> usize {
        match self {
            EngineWrapper::StandardUncompressed(e) => e.size(),
            EngineWrapper::ConcurrentUncompressed(e) => e.size(),
            EngineWrapper::StandardCompressed(e) => e.size(),
            EngineWrapper::ConcurrentCompressed(e) => e.size(),
            EngineWrapper::StandardART(e) => e.size(),
            EngineWrapper::ConcurrentART(e) => e.size(),
        }
    }

    fn stats(&self) -> EngineStats {
        match self {
            EngineWrapper::StandardUncompressed(e) => e.stats(),
            EngineWrapper::ConcurrentUncompressed(e) => e.stats(),
            EngineWrapper::StandardCompressed(e) => e.stats(),
            EngineWrapper::ConcurrentCompressed(e) => e.stats(),
            EngineWrapper::StandardART(e) => e.stats(),
            EngineWrapper::ConcurrentART(e) => e.stats(),
        }
    }
}

File: lib/go/engine.go
package radixip

import (
	"net"
	"runtime"
	"sync"
	"sync/atomic"
)

// STANDARD ENGINE

type StandardEngine struct {
	tree      RouteTree
	size      int64
	mu        sync.RWMutex
	statsData EngineStats
}

func NewStandardEngine(tree RouteTree) *StandardEngine {
	return &StandardEngine{
		tree:      tree,
		statsData: EngineStats{},
	}
}

func (e *StandardEngine) Insert(prefix *net.IPNet, metadata Metadata) error {
	ipnet := IpNetwork{IP: prefix.IP, Mask: prefix.Mask}
	isNew, err := e.tree.Insert(ipnet, metadata)
	if err != nil {
		return err
	}
	if isNew {
		atomic.AddInt64(&e.size, 1)
	}
	e.mu.Lock()
	e.statsData.Inserts++
	e.statsData.Size = int(atomic.LoadInt64(&e.size))
	e.mu.Unlock()
	return nil
}

func (e *StandardEngine) Lookup(ip net.IP) *Metadata {
	result := e.tree.Lookup(&ip)
	e.mu.Lock()
	e.statsData.Lookups++
	if result != nil {
		e.statsData.Hits++
	} else {
		e.statsData.Misses++
	}
	e.mu.Unlock()
	return result
}

func (e *StandardEngine) Remove(prefix *net.IPNet) *Metadata {
	ipnet := IpNetwork{IP: prefix.IP, Mask: prefix.Mask}
	removed := e.tree.Remove(&ipnet)
	if removed != nil {
		atomic.AddInt64(&e.size, -1)
		e.mu.Lock()
		e.statsData.Removals++
		e.statsData.Size = int(atomic.LoadInt64(&e.size))
		e.mu.Unlock()
	}
	return removed
}

func (e *StandardEngine) Contains(prefix *net.IPNet) bool {
	ipnet := IpNetwork{IP: prefix.IP, Mask: prefix.Mask}
	return e.tree.Contains(&ipnet)
}

func (e *StandardEngine) Clear() {
	e.tree.Clear()
	atomic.StoreInt64(&e.size, 0)
	e.mu.Lock()
	e.statsData.Size = 0
	e.mu.Unlock()
}

func (e *StandardEngine) Size() int {
	return int(atomic.LoadInt64(&e.size))
}

func (e *StandardEngine) Stats() *EngineStats {
	e.mu.RLock()
	defer e.mu.RUnlock()
	s := e.statsData
	s.Size = int(atomic.LoadInt64(&e.size))
	return &s
}

// SHARDED ENGINE
// throughput = num_shards * throughput_per_shard

type ShardedEngine struct {
	shards    []*StandardEngine
	numShards int
	maskBits  int
}

func NewShardedEngine(numShards int, nodeVariant NodeVariant) *ShardedEngine {
	return NewShardedEngineWithTree(numShards, func() RouteTree {
		return NewUncompressedTree(nodeVariant)
	})
}

// NewShardedEngineWithTree accepts a factory so each shard can own a separate tree instance.
func NewShardedEngineWithTree(numShards int, treeFn func() RouteTree) *ShardedEngine {
	shards := make([]*StandardEngine, numShards)
	for i := 0; i < numShards; i++ {
		shards[i] = NewStandardEngine(treeFn())
	}
	return &ShardedEngine{shards: shards, numShards: numShards, maskBits: 24}
}

func (e *ShardedEngine) getShard(ip *net.IP) int {
	var hash uint64
	switch {
	case ip.To4() != nil:
		ip4 := ip.To4()
		/*
			Thinking of it as reclaiming their "favourite sitting spots" inside a larger container perfectly captures exactly what the hardware is doing.
			To solidify your intuition,
			let's look closely at your shifting counts,
			because you have the concept 100% correct,
			you just had a tiny typo on the exact numbers for the 3rd and 4th octets:
			1st Octet: Starts at the bottom of its own small container.
			It shifts 24 slots left to sit at the very top of the 32-bit container.
			2nd Octet: Starts at the bottom. It shifts 16 slots left to sit right next to the 1st octet.
			3rd Octet: Starts at the bottom.
			It shifts 8 slots left (not 16) to slide into the third position.4th Octet: It shifts 0 slots (it doesn't shift at all, or shifts 8 less than the 3rd) because its "favourite spot" is already at the very bottom!

		*/
		hash = uint64(uint32(ip4[0])<<24 | uint32(ip4[1])<<16 | uint32(ip4[2])<<8 | uint32(ip4[3]))
		// Clear out the host bits based on the matching CIDR mask
		var mask uint32 = 0xFFFFFFFF
		if e.maskBits < 32 {
			mask = mask << (32 - e.maskBits)
		}
		// Modulo the masked network identifier
		/*
					When you perform hash & mask, you zero out (flush) the host bits at the end of the IP address.This operation works because a bitwise & (AND) acts like a filter:Any Bit & 1 = Any Bit (The bit stays exactly the same)Any Bit & 0 = 0 (The bit is completely flushed/cleared)Visual example with an IP addressLet's look at what happens when your /24 mask (which has eight 0s at the end) meets the IP address:textIP Address:   11000000 10101000 00000001 00110010  (192.168.1.50)
			Mask (&):     11111111 11111111 11111111 00000000  (255.255.255.0)
			-------------------------------------------------
			Result:       11000000 10101000 00000001 00000000  (192.168.1.0)
			                                         ^^^^^^^^
			                                    Flushed to zeros
			Use code with caution.Because the mask ends in all zeros, it forces the last 8 bits of the IP address to become 0, leaving you with the pure network prefix.
		*/
		maskedHash := hash & uint64(mask)

		return int(uint64(maskedHash) % uint64(e.numShards))
	default:
		// IPv6 Implementation (Using the same strategy to maintain performance)
		ip6 := ip.To16()
		high := uint64(ip6[0])<<56 | uint64(ip6[1])<<48 | uint64(ip6[2])<<40 | uint64(ip6[3])<<32 |
			uint64(ip6[4])<<24 | uint64(ip6[5])<<16 | uint64(ip6[6])<<8 | uint64(ip6[7])
		var mask6 uint64 = 0xFFFFFFFFFFFFFFFF
		if e.maskBits < 64 {
			mask6 = mask6 << (64 - e.maskBits)
		}
		hash = high & mask6
	}
	return int(hash % uint64(e.numShards))
}

func (e *ShardedEngine) Insert(prefix *net.IPNet, metadata Metadata) error {
	// Insert into all shards so any shard can serve lookups correctly
	for _, shard := range e.shards {
		if err := shard.Insert(prefix, metadata); err != nil {
			return err
		}
	}
	return nil
}

func (e *ShardedEngine) Lookup(ip net.IP) *Metadata {
	return e.shards[e.getShard(&ip)].Lookup(ip)
}

func (e *ShardedEngine) Remove(prefix *net.IPNet) *Metadata {
	var removed *Metadata
	for _, shard := range e.shards {
		if r := shard.Remove(prefix); removed == nil {
			removed = r
		}
	}
	return removed
}

func (e *ShardedEngine) Contains(prefix *net.IPNet) bool {
	if len(e.shards) > 0 {
		return e.shards[0].Contains(prefix)
	}
	return false
}

func (e *ShardedEngine) Clear() {
	for _, shard := range e.shards {
		shard.Clear()
	}
}

func (e *ShardedEngine) Size() int {
	if len(e.shards) > 0 {
		return e.shards[0].Size()
	}
	return 0
}

func (e *ShardedEngine) Stats() *EngineStats {
	var total EngineStats
	for _, shard := range e.shards {
		s := shard.Stats()
		total.Lookups += s.Lookups
		total.Hits += s.Hits
		total.Misses += s.Misses
	}
	if len(e.shards) > 0 {
		s := e.shards[0].Stats()
		total.Inserts = s.Inserts
		total.Removals = s.Removals
	}
	total.Size = e.Size()
	return &total
}

// ENGINE WRAPPER
// Single entry-point that dispatches to the right engine+tree combination.

type EngineWrapper struct {
	engine  RadixEngine
	variant EngineVariant
}

// NewEngineWrapper creates an engine using the UncompressedTree (default).
func NewEngineWrapper(variant EngineVariant, nodeVariant NodeVariant) *EngineWrapper {
	return NewEngineWrapperWithTree(variant, nodeVariant, false)
}

// NewEngineWrapperWithTree allows choosing compressed vs uncompressed tree at construction.
// compressed=true uses CompressedTree (Patricia); compressed=false uses UncompressedTree (bitwise trie).
func NewEngineWrapperWithTree(variant EngineVariant, nodeVariant NodeVariant, compressed bool) *EngineWrapper {
	treeFn := func() RouteTree {
		if compressed {
			return NewCompressedTree(nodeVariant)
		}
		return NewUncompressedTree(nodeVariant)
	}

	var engine RadixEngine
	switch variant {
	case EngineStandard:
		engine = NewStandardEngine(treeFn())
	case EngineConcurrent:
		engine = NewShardedEngineWithTree(16, treeFn)
	case EngineLockFree:
		// LockFree uses atomic nodes; uncompressed for simplicity
		engine = NewStandardEngine(NewUncompressedTree(NodeLockFree))
	case EngineAdaptive:
		cpus := runtime.NumCPU()
		if cpus > 4 {
			engine = NewShardedEngineWithTree(cpus*2, treeFn)
		} else {
			engine = NewStandardEngine(treeFn())
		}
	case EngineART:
		engine = NewARTEngineAdapter()
	default:
		engine = NewStandardEngine(treeFn())
	}

	return &EngineWrapper{engine: engine, variant: variant}
}

func (e *EngineWrapper) Insert(prefix *net.IPNet, metadata Metadata) error {
	return e.engine.Insert(prefix, metadata)
}

func (e *EngineWrapper) Lookup(ip net.IP) *Metadata {
	return e.engine.Lookup(ip)
}

func (e *EngineWrapper) Remove(prefix *net.IPNet) *Metadata {
	return e.engine.Remove(prefix)
}

func (e *EngineWrapper) Contains(prefix *net.IPNet) bool {
	return e.engine.Contains(prefix)
}

func (e *EngineWrapper) Clear() {
	e.engine.Clear()
}

func (e *EngineWrapper) Size() int {
	return e.engine.Size()
}

func (e *EngineWrapper) Stats() *EngineStats {
	stats := e.engine.Stats()
	return stats
}
